cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,487 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
+      val visitedBefore = coloring.get(next) match {
+        case Walk.NodeColor.Visited => true
+        case _ => false
+      }
+      if (visitedBefore) {
+        handleVisitedInQueue(next)
+        findNextNodeToVisit()
+      } else {
+        Some(next)
       }
     }
   }
 
-  private[graph] def depth(id: Int) = depthTracker.infoOfNode(id)
-  private def seen(id: Int) = depth(id).isDefined
-
-  def next = {
-    val curr = qu.dequeueInt()
-    val currDepth = depth(curr).get
-    val nd = getExistingNodeById(graph, curr)
-    log.ifTrace { "visiting %d, nbrCount=%d, maxNumEdges=%d, depth=%d".format(curr,
-      nd.neighborCount(dir), maxNumEdgesThresh.getOrElse(-1), currDepth) }
-    val newDepth = currDepth + 1
-    enqueueNeighbors(curr, newDepth)
-    nd
+  def handleVisitedInQueue(next: Int) {
+    queue.dequeueInt()
   }
 
-  def hasNext = if (qu.size == 0) {
-    Stats.incr("bfs_walk_request_exhausted_2nd_degree", 1)
-    false
-  } else {
-    maxDepth.isEmpty || depth(qu.firstInt()).get <= maxDepth.get
-  }
+  // can't limit number of nodes added to the queue, because we skip some nodes
+  override def limitAddedToQueue(nodes: Seq[Int]): Seq[Int] = nodes
+
+  override def shouldBeEnqueued(color: Walk.NodeColor.Color): Boolean = color != Walk.NodeColor.Visited
 }
 
-class DepthFirstTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
-    onlyOnce: Boolean)
-    extends DistanceTraverser[Int](graph, homeNodeIds, onlyOnce) {
 
-  // stack keeps a tuple (node, child# of higher node in stack). Hence, top node
-  // of stack always has child# = -1
-  case class NodeDesc(val id: Int, val childOff: Int)
+/**
+ * A trait to be mixed in to the DepthFirstTraverser that keeps track of visiting distance
+ * from `homeNodeIds`.
+ *
+ * Note that DepthTracker for the DFS case returns the depth of first seeing a particular
+ * node. It does not have to be the same as the depth at which the node is visited.
+ */
+trait PathLengthTracker extends DepthFirstTraverser {
+  private lazy val nextVisitDistanceTracker = new IntInfoKeeper(false)
+  private lazy val visitDistanceTracker = new IntInfoKeeper(true)
+
+  abstract override protected def enqueue(nodes: Seq[Int], from: Option[Int]): Unit = {
+    val fromDistance = from.flatMap(visitDistanceTracker.infoOfNode).getOrElse(-1)
+    nodes foreach { id =>
+      nextVisitDistanceTracker.recordInfo(id, fromDistance + 1)
+    }
+    super.enqueue(nodes, from)
+  }
 
-  private val stack = new ObjectArrayList[NodeDesc]
+  override def visitNode(node: Node) {
+    visitDistanceTracker.recordInfo(node.id, nextVisitDistanceTracker.infoOfNode(node).get)
+    super.visitNode(node)
+  }
 
-  init(-1)
+  def distance(id: Int) = visitDistanceTracker.infoOfNode(id)
+}
 
-  protected def storeForVisit(id: Int, info: Int) { stack.add(NodeDesc(id, info)) }
+/**
+ * Trait to be mixed in to the DFS traverser to add discovery time of a node.
+ */
+trait DiscoveryTimeTracker extends DepthFirstTraverser {
```

Comment:
shouldn't this (as well as FinishingTimeTracker) extend QueueBasedTraverser so that it can be used in BFS as well?